InteractionBase/src/vis/Menu.cpp

```diff
@@ -0,0 +1,245 @@
+#include "VisualizationBase/src/declarative/DynamicGridFormElement.h"
+#include "VisualizationBase/src/declarative/GridLayouter.h"
+#include "VisualizationBase/src/items/Text.h"
+#include "VisualizationBase/src/cursor/TextCursor.h"
+#include "QGraphicsColorizeEffect"
+#include "VisualizationBase/src/items/Item.h"
+#include "nodes/ViewSwitcherEntry.h"
+
+namespace Interaction {
+
+ITEM_COMMON_DEFINITIONS(Menu, "item")
+
+Menu::Menu(QVector<Model::Node*> selectableNodes,
+			Visualization::Item* target, StyleType* style, int nrOfColumns)
+	: Menu(arrange(selectableNodes, nrOfColumns), target, style)
+{
+}
+
+Menu::Menu(QVector<QVector<Model::Node *> > selectableNodes,
+			Visualization::Item *target, StyleType *style)
+	: Super(nullptr, style), target_(target), currentNodes_(selectableNodes)
+{
+	mousePosition_ = target->scene()->lastMouseHoverPosition();
+	selectedEffect_ = new QGraphicsColorizeEffect();
+	if (hasTextField())
+	{
+		textField_ = new Visualization::Text(this, &style->nameField(), "Enter name here");
+		textField_->setEditable(true);
+	}
+}
+
+Menu::~Menu()
+{
+	//Move the cursor to original owner
+	target_->moveCursor();
+	//Shouldn't be destroyed
+	target_ = nullptr;
+	//Destroyed by DeclarativeItem
+	focusedItem_ = nullptr;
+	SAFE_DELETE_ITEM(textField_);
+	//Destroyed when the focused item is destroyed
+	if (!focusedItem() || focusedItem()->graphicsEffect() != selectedEffect_)
+		SAFE_DELETE(selectedEffect_);
+	selectedEffect_ = nullptr;
+}
+
+void Menu::initializeForms()
+{
+	auto nodeGrid = (new Visualization::DynamicGridFormElement())
+			->setSpacing(5, 5)
+			->setMajorAxis(Visualization::GridLayouter::ColumnMajor)
+			->setNodesGetter([](Visualization::Item* v)
+				{ auto self = static_cast<I*>(v);
+				  return self->currentNodes(); });
+	//First form: only the nodes
+	addForm(nodeGrid);
+	//Second form: the nodes and a text field
+	addForm((new Visualization::GridLayoutFormElement())
+			->setSpacing(5, 5)->setMargins(10)
+			->put(0, 0, item<Visualization::Text>(&I::textField_,
+									[](I* v) -> const Visualization::TextStyle* { return &v->style()->nameField(); }))
+									//[](I* v) { return v->style()->nameField(); }))
+			->put(0, 1, nodeGrid));
+}
+
+int Menu::determineForm()
+{
+	return hasTextField() ? 1 : 0;
+}
+
+bool Menu::hasTextField() const
+{
+	return true;
+}
+
+void Menu::selectItem(Visualization::Item *item)
+{
+	if (focusedItem() && focusedItem() != textField())
+	{
+		focusedItem()->setGraphicsEffect(nullptr);
+		selectedEffect_ = new QGraphicsColorizeEffect();
+	}
+	if (item && item != textField())
+		item->setGraphicsEffect(selectedEffect_);
+	if (item == textField())
+	{
+		item->moveCursor();
+		item->correspondingSceneCursor<Visualization::TextCursor>()->selectAll();
+	}
+	else moveCursor(Visualization::Item::MoveOnPosition, QPoint(0, 0));
+	focusedItem_ = item;
+}
+
+bool Menu::executeFocused()
+{
+	if (hasTextField() && focusedItem() == textField())
```

Comment:
I think in the end we should end-up with something similar to the AutoCompletion:
- individual entries should know how to execute themselves and we should just call this method here
- we should do away with the text/selection distinction and just have menu entries.